xonsh/built_ins.py

```diff
@@ -317,322 +306,494 @@ def _redirect_io(streams, r, loc=None):
+        if callable(self.alias):
+            p = self.cls(self.alias, self.cmd, **kwargs)
+        else:
+            p = self._run_binary(kwargs)
+        p.last_in_pipeline = self.last_in_pipeline
+        p.captured_stdout = self.captured_stdout
+        p.captured_stderr = self.captured_stderr
+        return p
+
+    def _run_binary(self, kwargs):
+        try:
+            bufsize = 1
+            p = self.cls(self.cmd, bufsize=bufsize, **kwargs)
+        except PermissionError:
+            e = 'xonsh: subprocess mode: permission denied: {0}'
+            raise XonshError(e.format(self.cmd[0]))
+        except FileNotFoundError:
+            cmd0 = self.cmd[0]
+            e = 'xonsh: subprocess mode: command not found: {0}'.format(cmd0)
+            env = builtins.__xonsh_env__
+            sug = suggest_commands(cmd0, env, builtins.aliases)
+            if len(sug.strip()) > 0:
+                e += '\n' + suggest_commands(cmd0, env, builtins.aliases)
+            raise XonshError(e)
+        return p
+
+    def prep_env(self, kwargs):
+        """Prepares the environment to use in the subprocess."""
+        denv = builtins.__xonsh_env__.detype()
+        if ON_WINDOWS:
+            # Over write prompt variable as xonsh's $PROMPT does
+            # not make much sense for other subprocs
+            denv['PROMPT'] = '$P$G'
+        kwargs['env'] = denv
+
+    def prep_preexec_fn(self, kwargs, pipeline_group=None):
+        """Prepares the 'preexec_fn' keyword argument"""
+        if not (ON_POSIX and self.cls is subprocess.Popen):
+            return
+        if pipeline_group is None:
+            xonsh_preexec_fn = no_pg_xonsh_preexec_fn
+        else:
+            def xonsh_preexec_fn():
+                """Preexec function bound to a pipeline group."""
+                os.setpgid(0, pipeline_group)
+                signal.signal(signal.SIGTSTP, default_signal_pauser)
+        kwargs['preexec_fn'] = xonsh_preexec_fn
+
+    #
+    # Building methods
+    #
+
+    @classmethod
+    def build(kls, cmd, *, cls=subprocess.Popen, **kwargs):
+        """Creates an instance of the subprocess command, with any
+        modifcations and adjustments based on the actual cmd that
+        was recieved.
+        """
+        # modifications that do not alter cmds may come before creating instance
+        spec = kls(cmd, cls=cls, **kwargs)
+        # modifications that alter cmds must come after creating instance
+        spec.redirect_leading()
+        spec.redirect_trailing()
+        spec.resolve_alias()
+        spec.resolve_binary_loc()
+        spec.resolve_auto_cd()
+        spec.resolve_executable_commands()
+        spec.resolve_alias_cls()
+        return spec
+
+    def redirect_leading(self):
+        """Manage leading redirects such as with '< input.txt COMMAND'. """
+        while len(self.cmd) >= 3 and self.cmd[0] == '<':
+            self.stdin = safe_open(seld.cmd[1], 'r')
```

Comment:
Which part?  `safe_open()` is a function at module level scope. See https://github.com/xonsh/xonsh/pull/1758/files/965d2aa2569312cee9fcb3440450e35a81c56cc1#diff-32a5e51471ce3482f01745a4592320c1R267